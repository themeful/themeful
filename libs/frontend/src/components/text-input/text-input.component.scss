:host {
  display: block;
}

.text-input {
  color: var(--tf-font-color);

  &__label {
    font: var(--tf-small-font);
    width: 100%;
  }

  &__hint {
    font: var(--tf-small-font);
    display: block;
    min-height: 1rem;
    width: 100%;
    color: #ff6464;
  }

  &__input {
    font: var(--tf-default-font);
    width: 100%;
    color: var(--tf-font-color);
    border: 0;
    background-color: rgba(128, 128, 128, 0.1);
    border-bottom: 1px solid currentColor;

    &:focus {
      outline: 0 !important;
      border-color: #325af0;
    }
  }

  &--error &__input {
    border-color: #ff6464;
  }

  &--valid &__input {
    border-color: #23c223;
  }
}
