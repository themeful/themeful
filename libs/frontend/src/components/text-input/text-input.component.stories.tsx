import { h } from '@stencil/core'
import { action } from '@storybook/addon-actions'

export default {
  title: 'Components/Text Input',
  args: {
    args: {
      type: 'text',
      label: 'the label',
      value: 'My Text',
    },
    validation: (value) => (Number(value) > 0 ? null : 'Please enter a number'),
  },
}

export const textInput = ({ args, validation }) => {
  return (
    <div>
      <div class="header">
        <h1>Alias Token Form</h1>
      </div>
      <div class="content">
        <div style={{ position: 'relative', width: '300px', padding: '16px' }}>
          <tf-text-input {...args} onInputChange={action('change')} />
        </div>
        <div
          class="tf-light"
          style={{ position: 'relative', width: '300px', background: '#fff', padding: '16px' }}
        >
          <tf-text-input {...args} onInputChange={action('change')} />
        </div>
        <div style={{ position: 'relative', width: '300px', padding: '16px' }}>
          <tf-text-input {...args} min-length="10" onInputChange={action('change')} />
        </div>
        <div style={{ position: 'relative', width: '300px', padding: '16px' }}>
          <tf-text-input {...args} validation={validation} onInputChange={action('change')} />
        </div>
      </div>
    </div>
  )
}
