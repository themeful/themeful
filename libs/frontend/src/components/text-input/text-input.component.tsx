import { Component, Element, Event, EventEmitter, h, Prop, State, Watch } from '@stencil/core'
@Component({
  tag: 'tf-text-input',
  styleUrl: 'text-input.component.scss',
})
export class TextInputComponent {
  @Element() el: HTMLTfTextInputElement

  /** Input type */
  @Prop() type = 'text'

  /** Input label */
  @Prop() label: string

  /** Required input */
  @Prop() required = false

  /** Min input */
  @Prop() minLength: number

  /** Max input */
  @Prop() maxLength: number

  /** Input value */
  @Prop({ mutable: true }) value: string

  /** validation function */
  @Prop() validation: (string) => string | null

  /** Input Event */
  @Event() inputChange: EventEmitter

  @State() touched = false
  @State() changed = false
  @State() valid: boolean
  @State() errorHint = ''

  @Watch('value')
  valueChanged() {
    const inputEl = this.el.querySelector('input')
    if (inputEl.value !== this.value) {
      inputEl.value = this.value
    }
  }

  private inputChanged = (event: Event) => {
    const val = event.target && (event.target as HTMLInputElement).value
    this.changed = true
    this.value = val
    this.inputChange.emit(this.value)
    if (!this.valid) {
      this.validate()
    }
  }

  private blur = (): void => {
    this.validate()
    this.touched = true
  }

  private validate = (): void => {
    let state = true
    let error = ''
    if (this.validation !== undefined) {
      const result = this.validation(this.value)
      state = result === null
      error = state ? '' : result
    }
    if (state && this.required) {
      state = this.value !== ''
      error = `This value is required`
    }
    if (state && this.minLength !== undefined) {
      state = this.value?.length >= this.minLength
      error = state ? '' : `The min length is ${this.minLength}`
    }
    if (state && this.maxLength !== undefined) {
      state = this.value?.length <= this.maxLength
      error = state ? '' : `The max length is ${this.maxLength}`
    }
    this.errorHint = error
    this.valid = state
  }

  render(): HTMLTfTextInputElement {
    return (
      <label
        class={`text-input${this.valid && this.touched ? ' text-input--valid' : ''}${
          !this.valid && this.touched ? ' text-input--error' : ''
        }`}
      >
        <span class="text-input__label">{this.label}</span>
        <input
          class="text-input__input"
          value={this.value}
          type={this.type}
          onInput={this.inputChanged}
          onBlur={this.blur}
        ></input>
        <p class="text-input__hint">{this.errorHint}</p>
      </label>
    )
  }
}
