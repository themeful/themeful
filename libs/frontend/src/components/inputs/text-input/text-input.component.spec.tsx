import { h } from '@stencil/core'
import { newSpecPage } from '@stencil/core/testing'
import { TextInputComponent } from './text-input.component'

describe('TextInputComponent', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [TextInputComponent],
      autoApplyChanges: true,
      template: () => <tf-text-input />,
    })

    expect(page.root).toMatchSnapshot()
  })

  it('renders with values', async () => {
    const page = await newSpecPage({
      components: [TextInputComponent],
      autoApplyChanges: true,
      template: () => <tf-text-input label="Some Label" value="Test Text" />,
    })

    expect(page.root).toMatchSnapshot()
  })
})
