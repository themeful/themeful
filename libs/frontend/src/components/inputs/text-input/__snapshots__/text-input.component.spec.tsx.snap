// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInputComponent renders 1`] = `
<tf-text-input>
  <mock:shadow-root>
    <label class="text-input">
      <span class="text-input__label"></span>
      <input class="text-input__input" type="text">
      <p class="text-input__hint"></p>
    </label>
  </mock:shadow-root>
</tf-text-input>
`;

exports[`TextInputComponent renders with values 1`] = `
<tf-text-input>
  <mock:shadow-root>
    <label class="text-input">
      <span class="text-input__label">
        Some Label
      </span>
      <input class="text-input__input" type="text" value="Test Text">
      <p class="text-input__hint"></p>
    </label>
  </mock:shadow-root>
</tf-text-input>
`;
